Discard DataStorm partial updates that have no base value (#6083)
zeroc-ice/ice: changelog.d/datastorm/partial-update-no-base.md

- Fixed a bug in DataStorm where a partial update for a key with no current value—no full value was written for the
  key, or the key was removed—was resolved against an invalid base value: applications using class-typed values
  typically crashed, and with other value types the update was silently applied to a default-constructed value,
  resurrecting removed keys. A remove now clears the key's value, and a partial update for a key with no value is
  discarded, by both writers and readers.
- Fixed a bug in DataStorm where a remove sample's value was indeterminate for scalar value types: `getValue()` on

>   the remove sample returned arbitrary data instead of a default-constructed value.

Reviewer: Two nits on this fragment:

- `changelog.d/README.md` maps `datastorm/` to the "Ice Service Changes > DataStorm" section, and the collation script generates that heading — so "in DataStorm" is redundant here.
- The em dashes are unspaced, so `value—no full value` reads as a hyphenated compound rather than an aside. Every other em dash in `changelog.d/` is spaced (`js/5876.md`, `service-icestorm/5848.md`, `datastorm/reader-partial-update-robustness.md`). Since the aside also splits the subject from its verb, it's cleaner as its own sentence.

```suggestion
- Fixed a bug where a partial update for a key with no current value was resolved against an invalid base value:
  applications using class-typed values typically crashed, and with other value types the update was silently
  applied to a default-constructed value, resurrecting removed keys. A key has no current value when no full value
  was written for it yet, or when it was removed. A remove now clears the key's value, and a partial update for a
  key with no value is discarded, by both writers and readers.
- Fixed a bug where a remove sample's value was indeterminate for scalar value types: `getValue()` on the remove
  sample returned arbitrary data instead of a default-constructed value.
```

Note that 7 of the 10 existing `changelog.d/datastorm/` fragments do open with "Fixed a bug in DataStorm", so this makes the new one the odd one out until the others are swept. Happy to leave it as-is if you'd rather keep the section internally consistent for now.

Author: 🤖 **Claude:** Em dashes spaced/split (the first entry was reworded for the throw regardless). Kept the "in DataStorm" opener for consistency with the 7/10 existing `datastorm/` fragments — happy to sweep them all together separately. (870c7b55a6 / 3effe09b91)

>   the remove sample returned arbitrary data instead of a default-constructed value.

Reviewer 2: "a remove sample's value was indeterminate" is written from the implementation's point of view — "remove sample" isn't a term a user necessarily knows. The user-visible contract lives on `getValue()`, whose doc already promises a default value when the sample event is Remove, so lead with that:

```suggestion
- Fixed a bug in DataStorm where calling `getValue()` on a sample carrying the `Remove` event returned indeterminate
  data instead of the documented default-constructed value, when the topic's value type is a scalar type.
```

(Independent of @externl's suggestion above — whichever lands second needs a trivial rebase.)

Author: 🤖 **Claude:** Applied your wording (3effe09b91).

>   data instead of the documented default-constructed value, when the topic's value type is a scalar type.

Reviewer 2: Suggestion: shorter sentences, drop "in DataStorm" (implied by the section), and fold the scalar-type condition into "could".

```suggestion
- Fixed the handling of partial updates for keys with no value. A key has no value when it was removed, or when no
  full value was written for it yet. Previously, such a partial update crashed applications using class-typed
  values; with other value types, it silently resurrected removed keys. Now a writer throws `std::logic_error`
  when publishing such an update, and a reader discards incoming ones.
- Fixed a bug where calling `getValue()` on a `Remove` sample could return indeterminate data instead of the
  documented default value.
```
